src/Database/PostgreSQL/PQTypes/Internal/Connection.hs

```diff
@@ -170,23 +172,59 @@ poolSource cs numStripes idleTime maxResources = do
 -- Useful if one wants to implement custom connection source.
 connect :: ConnectionSettings -> IO Connection
 connect ConnectionSettings{..} = do
-  fconn <- BS.useAsCString (T.encodeUtf8 csConnInfo) c_PQconnectdb
+  fconn <- BS.useAsCString (T.encodeUtf8 csConnInfo) openConnection
   withForeignPtr fconn $ \connPtr -> do
     conn <- peek connPtr
     status <- c_PQstatus conn
     when (status /= c_CONNECTION_OK) $
-      throwLibPQError conn "connect"
+      throwLibPQError conn fname
     F.forM_ csClientEncoding $ \enc -> do
       res <- BS.useAsCString (T.encodeUtf8 enc) (c_PQsetClientEncoding conn)
       when (res == -1) $
-        throwLibPQError conn "connect"
+        throwLibPQError conn fname
     c_PQinitTypes conn
     registerComposites conn csComposites
     Connection <$> newMVar (Just ConnectionData {
       cdFrgnPtr = fconn
     , cdPtr     = conn
     , cdStats   = initialStats
     })
+  where
+    fname = "connect"
+
+    openConnection :: CString -> IO (ForeignPtr (Ptr PGconn))
+    openConnection conninfo = E.mask $ \restore -> do
+      -- We want to use non-blocking C functions to be able to observe
+      -- incoming asynchronous exceptions, hence we don't use
+      -- PQconnectdb here.
+      conn <- c_PQconnectStart conninfo
+      when (conn == nullPtr) $
+        throwError "PQconnectStart returned a null pointer"
+      -- Work around a bug in GHC that causes foreign pointer
```

Comment:
This bug appears to be fixed, BTW.

Reply:
Yes, but I think it was fixed in 8.0.1, so earlier versions are affected.